### PR TITLE
[XLA] Enhancements to algebraic_simplifier

### PR DESCRIPTION
**Pull request by AyanmoI**
Motivation: algebraic simplification => rsqrt(Inv(pow^2(rsqrt(pow^-2(x))))) = |x| => x
This pattern found in one of the consumers of BatchNormForwardTraining. The producer of this pattern is actually the inverse sqrt of variance+epsilon which is always positive. Encoded this condition along with a few others when handling the transform |x| = x.

I had been working on mobilenet performance on the batchnorm issue(enabling fp16 cudnn batchnorm thunk to half kernels) which lead me to this optimization opportunity. With this change the graph is cleaner and also avoids having a multioutput fused kernel. There are quite a few instances of this pattern occurring in the graph. The gains with this optimization is ~1-2% for mobilenet.

This PR adds cases to HandleRsqrt, HandleSqrt, HandleMultiply (pow^2 gets converted to multiply) and HandleAbs. HandleAbs does the transform |x|=x when x is positive.

Screen Shot 2019-11-14 at 11 42 03 AM

### CODE REVIEW

**Comment by sanjoy** [comment]
Sorry, @akuegel is OOO.  Assigning to @bixia1

File: tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc [hunk 1]
```diff
@@ -131,6 +132,8 @@ Status NVPTXCompiler::OptimizeHloConvolutionCanonicalization(
                                           /*allow_mixed_precision=*/false);
 
     AlgebraicSimplifierOptions options;
+    options.set_cudnn_batchnorm_forward_training_metadata(
+        kCudnnBatchNormForwardTrainingCallTarget);
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc]: This custom call is created by Visitor::HandleBatchNormTraining. Can we run algebraic_simplifier before that and use the fact "batchnorm HLO returns {output, mean, variance} and variance is non-negative" to perform the optimization? In this case, I think you may not even need to rely on AlgebraicSimplifierOptions turn on/off the optimization.

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc]: I have thought about this but I couldn't not come to the conclusion that the pattern (motivation pattern mentioned in the description) emerges before the rewriter pass. What I can do is add another case to `IsNonNegative` (`IsPositive`) that handles batchnorm HLO. That seems to be a natural extension anyway.

Comment (bixia1) [reply tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc]: Can you provide some detail on how the original code and the code generated by Visitor::HandleBatchNormTraining interact and produce the pattern? What I can tell is the transformation replaces a use of the x (batch norm result with index 2) with this function of y (custom call result with index 2):  power(y, -2) - epsilon.

**Comment by AyanmoI** [comment]
Hi @bixia1 ... thanks for pointing out the NAN handling characteristics in Tensorflow. I have made modifications accordingly. Please note that now I have 2 functions: `IsNonNegative` and `IsPositive`. `HandleAbs` uses `IsNonNegative` to make decisions while all other transforms verify that the operand is strictly positive (using `IsPositive`). Also, note that although the transform rsqrt(1/x)-> sqrt(x) would work for any non-zero value, I chose to be more conservative since isolating just non-zero value would be trickier than just isolating positive values and this conservative constraint would still work for our current use case. I have assumed that rsqrt(variance+epsilon) always positive to be the base case (ground truth). More cases can be added as when required.

File: tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/nvptx_compiler.cc]: Before `cudnn_batchnorm_rewiter`, the pattern that exist is `Inv([rsqrt(x+c)*rsqrt(x+c)])-c` which gets transformed to the one handled in this PR. This can also be simplified to `x` but might prove to be non-trivial. I can check if this pattern gets eliminated by existing transforms+the ones I've added here. However, do you think we need to add this case as well?

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc [hunk 2]
```diff
@@ -5809,5 +5810,235 @@ TEST_F(AlgebraicSimplifierTest, SliceOfConcat) {
               GmockMatch(m::Parameter(1)));
 }
 
+TEST_F(AlgebraicSimplifierTest, SqrtOfSelfMultiply) {
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Would you please simplify the names used in the HLO instructions for the tests you added, to make it more readable? See CopyReshape for an example.

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc [hunk 3]
```diff
@@ -5809,5 +5810,235 @@ TEST_F(AlgebraicSimplifierTest, SliceOfConcat) {
...
+      ROOT %sqrt = f32[32]{0} sqrt(f32[32]{0} %multiply)
+    }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(kModuleStr));
+  ASSERT_TRUE(AlgebraicSimplifier(default_options_).Run(m.get()).ValueOrDie());
+  EXPECT_THAT(m->entry_computation()->root_instruction(),
+              GmockMatch(m::Abs(m::Parameter(0))));
+}
+
+TEST_F(AlgebraicSimplifierTest, RsqrtOfRPower) {
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: For a few tests you added that use the meta data, I think we need to add corresponding negative tests. The negative tests may not set the meta data, it may also just use a non-batch-norm related HLO value.

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: I agree. Will do that.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 4]
```diff
@@ -3283,6 +3360,31 @@ Status AlgebraicSimplifierVisitor::HandleSlice(HloInstruction* slice) {
   return Status::OK();
 }
 
+Status AlgebraicSimplifierVisitor::HandleRsqrt(HloInstruction* rsqrt) {
+  VLOG(10) << "trying transform [rsqrt(Pow(A, -2)) => |A|] "
+           << rsqrt->ToString();
+  HloInstruction* rsqrt_operand = rsqrt->mutable_operand(0);
```
[the remark this answers is not in the document]
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: `rsqrt_operand` here is not `A` but `Pow`. Do you want me to change `rsqrt_operand` to something like `pow`? But in that case, it might be misleading since `rsqrt`'s operand may or may not be `Pow`. Sorry if I misunderstood your comment.

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc (hunk 3 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Btw do you want negative tests for all these tests? or would adding one or two cases suffice?

Comment (bixia1) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: I think two is good enough.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bixia1) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: I mis-read the code, rsqrt_operand is fine.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 5]
```diff
@@ -2128,14 +2192,27 @@ Status AlgebraicSimplifierVisitor::HandleMultiply(HloInstruction* multiply) {
...
     return ReplaceWithNewInstruction(
         multiply,
         HloInstruction::CreateUnary(multiply->shape(), HloOpcode::kExp, add));
   }
+
+  VLOG(10) << "trying transform [rsqrt(A) * rsqrt(A) => 1/A] "
+           << multiply->ToString();
+  HloInstruction* b;
+  if (Match(multiply, m::Multiply(m::Rsqrt(m::Op(&b)), m::Rsqrt(m::Op(&b)))) &&
+      IsPositive(multiply->operand(0), options_)) {
```
[the remark this answers is not in the document]
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: Done

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc (hunk 2 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Done

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc (hunk 3 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: The negative tests I added may a too detailed. Please let me know if this is what you expect or would just asserting that AlgebraicSimplifier change hasn't kicked in is enough. (`ASSERT_FALSE(AlgebraicSimplifier(default_options_).Run(m.get()).ValueOrDie());`)

File: tensorflow/compiler/xla/service/algebraic_simplifier.h [hunk 6]
```diff
@@ -99,14 +99,27 @@ class AlgebraicSimplifierOptions {
...
+    Metadata() {}
+  };
   ReshapeIsBitcastCallback reshape_is_bitcast_callback_;
   bool is_layout_sensitive_{false};
   bool enable_dot_strength_reduction_{true};
   bool enable_dot_to_multiply_rewrite_{true};
   bool enable_conv_simplification_{true};
   bool enable_window_reduce_to_reduce_replacement_{true};
   int64 very_small_gather_size_{4};
+  Metadata metadata_;
```
[the remark this answers is not in the document]
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.h]: Done.

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc [hunk 7]
```diff
@@ -5809,5 +5810,354 @@ TEST_F(AlgebraicSimplifierTest, SliceOfConcat) {
...
+      get-tuple-element.2 = f32[32]{0} get-tuple-element(custom-call.1), index=1
+      get-tuple-element = f32[32]{0} get-tuple-element(custom-call.1), index=2
+      abs = f32[32]{0} abs(get-tuple-element)
+      ROOT %tuple = (f32[128,32,2,112]{3,2,1,0}, f32[32]{0}, f32[32]{0}) tuple(get-tuple-element.1, get-tuple-element.2, abs)
+    }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(kModuleStr));
+  default_options_.set_cudnn_batchnorm_forward_training_metadata(
+      "__cudnn$batchNormalizationForwardInference");
+  ASSERT_FALSE(AlgebraicSimplifier(default_options_).Run(m.get()).ValueOrDie());
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: I think ASSERT_FALSE is enough, no need to check for abs.

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc [hunk 8]
```diff
@@ -5809,5 +5810,354 @@ TEST_F(AlgebraicSimplifierTest, SliceOfConcat) {
...
+  default_options_.set_cudnn_batchnorm_forward_training_metadata(
+      "__cudnn$batchNormalizationForwardTraining");
+  ASSERT_TRUE(AlgebraicSimplifier(default_options_).Run(m.get()).ValueOrDie());
+  // Expected transformation: rsqrt(power(gte.2,-2)) -> abs(gte.2)
+  auto computation = m->entry_computation();
+  auto root = computation->root_instruction();
+  EXPECT_EQ(root->opcode(), HloOpcode::kTuple);
+  bool found_power = false;
+  bool found_rsqrt = false;
+  for (HloInstruction* inst : computation->instructions()) {
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Can you use HloTestBase::FindInstruction to find the instruction? Or you can add a routine to HloTestBase to find instruction by opcode.

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc (hunk 3 above)
Comment (bixia1) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Checking the pass return false is good enough here.

File: tensorflow/compiler/xla/service/algebraic_simplifier_test.cc (hunk 8 above)
Comment (bixia1) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Sorry, I don't see any response to this comment. There are other similar code in the following test cases. Can you respond to this?

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: Sorry.. I was out so couldn’t reply. I have made the changes and will push it today.

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier_test.cc]: @bixia1 I've pushed the change...is this what u had in mind? Please let me know.

File: tensorflow/compiler/xla/tests/hlo_test_base.h [hunk 9]
```diff
@@ -274,6 +274,7 @@ class HloTestBase : public ::testing::Test {
   // inspect a particular computation or instruction.
   HloComputation* FindComputation(HloModule* module, absl::string_view name);
   HloInstruction* FindInstruction(HloModule* module, absl::string_view name);
+  HloInstruction* FindInstruction(HloModule* module, HloOpcode opcode);
```
Comment (bixia1) [inline tensorflow/compiler/xla/tests/hlo_test_base.h]: The comment before the group is for finding the instruction with the given name. Can you modify the comment to including finding the instruction with the given opcode?

**Review by bixia1 — APPROVED** [review]

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 10]
```diff
@@ -80,6 +80,51 @@ bool IsAll(const HloInstruction* op, int8 value) {
...
+    default:
+      return false;
+  }
+}
+
+bool IsNonNegative(const HloInstruction* hlo,
+                   const AlgebraicSimplifierOptions& options) {
+  switch (hlo->opcode()) {
+    case HloOpcode::kMultiply: {
+      return hlo->operand(0) == hlo->operand(1);
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: For kMultiply, we need to make sure that the operand is not a complex number.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 11]
```diff
@@ -80,6 +80,51 @@ bool IsAll(const HloInstruction* op, int8 value) {
...
+          const auto& target = gte_operand->custom_call_target();
+          return target ==
+                     options.get_cudnn_batchnorm_forward_training_metadata() &&
+                 hlo->tuple_index() == 2;
+        }
+        default:
+          return false;
+      }
+    }
+    case HloOpcode::kPower:
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: For kPower, we need to make sure that the second operand is not a complex number, right?

**Review by bixia1 — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc (hunk 11 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: yea...looks like TF 2.0 allows complex for both operands. Not sure abt the semantics of TF 1.0 though. Anyway, I'll mk the change.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc (hunk 10 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: Infact, in that this case we should handle all cases when the operands are complex i.e, have an assertion that none of the operands can be complex type since `IsPositive` or `IsNonNegative` is not designed to used of Complex dtypes. IMO it would be better to have assertions to not allow complex.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc (hunk 11 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: Even operand(0) should not be complex. Only a pure imaginary number can generate a real number (that too negative). So it doesn't make sense to allow any operand to be complex.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc (hunk 10 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: Just to clarify, would you prefer an assertion when `IsPositive` or `IsNonNegative` is called with complex args or return false?

Comment (bixia1) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: It can just return false.

**Comment by AyanmoI** [comment]
Sorry for the late commit..I was OOO.

File: tensorflow/compiler/xla/service/algebraic_simplifier.cc [hunk 12]
```diff
@@ -80,6 +80,68 @@ bool IsAll(const HloInstruction* op, int8 value) {
...
+    }
+    default:
+      return false;
+  }
+}
+
+bool IsNonNegative(const HloInstruction* hlo,
+                   const AlgebraicSimplifierOptions& options) {
+  // Utility only handles real types.
+  if (IsAnyOperandComplex(hlo)) {
```
Comment (bixia1) [inline tensorflow/compiler/xla/service/algebraic_simplifier.cc]: abs(x) is always NonNegative, right?

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: Does abs(x) functionally produce a modulus value in TF? How is abs of cmplx handled in xla? I wasn’t sure so stuck to handling real values only.

Comment (bixia1) [reply tensorflow/compiler/xla/service/algebraic_simplifier.cc]: I think so. You can check the implementation of  ElementalIrEmitter::EmitComplexAbs.

**Review by bixia1 — APPROVED** [review]
